rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,970 @@
+  "kind": "entityType",
+  "$id": "https://blockprotocol.org/@alice/entity-type/region/v/1",
+  "type": "object",
+  "title": "Region",
+  "allOf": [
+    { "$ref": "https://blockprotocol.org/@alice/entity-type/country/v/1" }
+  ],
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/blurb": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/blurb/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/blurb"]
+}
+```
+
+The second, cyclic version of the `Country` entity type
+
+```json
+{
+  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
+  "kind": "entityType",
+  // Because of versioning, we cannot change this version to /v/1 and create a "proper" cycle.
+  "$id": "https://blockprotocol.org/@alice/entity-type/country/v/2",
+  "type": "object",
+  "title": "Country",
+  "allOf": [
+    { "$ref": "https://blockprotocol.org/@alice/entity-type/region/v/1" }
+  ],
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/location": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/location/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/name"]
+}
+```
+
+This sort of type hierarchy should _not_ be encouraged within the type extension system, as the circular dependencies make types difficult to reason about. While the type hierarchy might be completely valid (as it would be in this case), we should safeguard users from making redundant type structures that look like the above.
+
+In this specific contrived example, creating a new entity type based on `Region` instead of a new version of `Country` might even encode semantic meaning better than re-defining `Country`.
+
+## Multiple supertypes - checking compatibility
+
+As described in the [Guide-level explanation](#multiple-supertypes), when extending multiple entity types, they must be able to coexist in a "compatible manner", which means that the entity types' `properties` and `links` comply with the following:
+
+- For each property (base URL on the top level of `properties`) that exists in multiple entity types:
+
+  - all entity types refer to the same versioned URL of the property type **or** compatible versions of the property type (through the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility) definition of 'compatible')
+  - none of the entity types define the property type as an array **or** all define the property type as an array with compatible cardinality constraints.
+
+- For each link (versioned URL on the top level of `links`) that exists in multiple entity types:
+
+  - none of the entity types define the link as an array **or** all define the link as an array with compatible cardinality constraints (and same `order` value).
```

Comment:
All links are implicitly captured as "arrays" now. We just need to enforce compatible cardinality constraints, **destination** constraints, and ordering.

Compatible destination constraints is less clear to me.. We should be specific about this, I'm guessing we have to check compatibility of the union of link destinations that are allowed, where compatibility between entity types is effectively just accounting for inheritance? Or do we also need to consider versioning here 🙃

Reply:
This is now redundant in the reworked document as we don't require type extension to result in something satisfiable.